### PR TITLE
Port comparison to C++

### PR DESCRIPTION
**Pull request by wilhuff**
This reimplements our comparison functions as C++ Comparators and then provides compatibility shims for interoperating with existing Objective-C usage.

A few specialized comparators aren't suitable for porting but only have a single usage (e.g. `CompareBytes` for comparing `NSData *` instances). In these cases I've moved them into the caller.

### CODE REVIEW

File: Firestore/core/src/firebase/firestore/util/comparison.h [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+ * strongly-ordered values (i.e. where comparison between values always yields
+ * less-than, equal-to, or greater-than).
+ *
+ * This is equivalent to:
+ *
+ *   * NSComparisonResult from the iOS/macOS Foundation framework.
+ *   * std::strong_ordering from C++20
+ *
+ * The values of the constants are specifically chosen so as to make casting
+ * between this type and NSComparisonResult possible.
```
Comment (rsgowman) [inline Firestore/core/src/firebase/firestore/util/comparison.h]: That might be worth a quick test.
assert(nscomparison.ascending == comparisonresult.ascending);
...

Also along the same lines:
assert(ascending * -1 == decending);

Optional.

File: Firestore/core/test/firebase/firestore/util/comparison_test.cc [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace util {
+
+#define ASSERT_SAME(comparison)                    \
+  do {                                             \
+    ASSERT_EQ(ComparisonResult::Same, comparison); \
+  } while (0)
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/comparison_test.cc]: This construct allows a macro to be called as it it were a function in any context where you can call a function, even bone-headed code like this:

```c
if (condition)
  CALL1();
else
  CALL2();
```
You do this essentially for the same reason we require if/else constructs to always use braces.

File: Firestore/core/test/firebase/firestore/util/comparison_test.cc [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+#define ASSERT_DESCENDING(comparison)                    \
+  do {                                                   \
+    ASSERT_EQ(ComparisonResult::Descending, comparison); \
+  } while (0)
+
+TEST(Comparison, StringCompare) {
+  ASSERT_ASCENDING(Compare<absl::string_view>("a", "b"));
+  ASSERT_ASCENDING(Compare<absl::string_view>("a", "aa"));
+
```
[the remark this answers is not in the document]
Comment (wilhuff) [reply Firestore/core/test/firebase/firestore/util/comparison_test.cc]: Done.

File: Firestore/core/src/firebase/firestore/util/comparison.cc [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+    return true;
+  } else if (left >= right) {
+    return false;
+  } else {
+    // One or both left and right is NaN.
+    return isnan(left) && !isnan(right);
+  }
+}
+
+static constexpr double LONG_MIN_VALUE_AS_DOUBLE =
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/util/comparison.cc]: Yup, my use of "usually" was wrong.

File: Firestore/core/src/firebase/firestore/util/comparison.h [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+ * eventually be removed.
+ */
+template <typename T>
+inline NSComparisonResult WrapCompare(const T& left, const T& right) {
+  return static_cast<NSComparisonResult>(Compare<T>(left, right));
+}
+#endif
+
+/** Compares a double and an int64_t. */
+ComparisonResult CompareMixedNumber(double doubleValue, int64_t longValue);
```
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/util/comparison.h]: sgtm

File: Firestore/core/test/firebase/firestore/util/comparison_test.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (rsgowman) [reply Firestore/core/test/firebase/firestore/util/comparison_test.cc]: I hate macros. :(

**Pull request merged**